[export] disable_forced_specializations

pianpwk opened the pull request:
Summary:
By default, some inferred dynamic shapes guards/constraints that are not expressible with the current dynamic shapes language will lead to specialization to the concrete input values provided. If disable_forced_specializations is set to True, we will not specialize, and will not perform runtime checks on such produced guards. Instead, we allow the user to specify arbitrary shapes, and fail during runtime if the inputs are invalid. Constraints expressible with the language (e.g. ranges, linear derived dims) will still be enforced, and behavior for all other guards remains the same.

Cases where we typically specialize are reshapes:
```
x: [4, 6]  # [s0, s1]
x = x.reshape([x.shape[0] - 1, -1])
# this emits a guard Mod(s0*s1, s0-1) = 0, we specialize on s0=4, s1=6

x: [4, 6], y: [24]  # [s0, s1], [s2]
x = x.reshape([-1]) + y
# this emits a guard s0*s1 = s2, we specialize on s0=4, s1=6, s2=24
```

For now only applicable for non-strict mode (need to figure out how to pass this flag into dynamo's call of produce_guards).

Test Plan: Added test case that checks compilation, runtime, and suggested fixes behavior.

Differential Revision: D56361177


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1709,7 +1709,12 @@ def _specialize_divisor_symbols(self):
-    def solve(self, disable_congruences=True, disable_equivalences=True):
+    def solve(
+        self,
+        disable_congruences=True,
```
Too many flags, I'd get rid of the two previous ones and replace it with the new one.

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -1158,7 +1158,7 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
             ),
         ):
-            torch.export.export(m, (a,), dynamic_shapes=dynamic_shapes)
+            ep = torch.export.export(m, (a,), dynamic_shapes=dynamic_shapes)
```
are you using `ep`?

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -4480,6 +4480,144 @@ def forward(self, x, y, div="floor"):
+            r".*Specializations unexpectedly required(.*\n)*"
+            r"Suggested fixes:(.*\n)*"
+            r".*dy = Dim.*(.*\n)*"
+            r".*dw0 = None.*# 3(.*\n)*"
```
We should fix this suggestion separately, because I believe you can just say `3` now instead of `None`?

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -4480,6 +4480,144 @@ def forward(self, x, y, div="floor"):
+                inputs,
+                dynamic_shapes=dynamic_shapes,
+                strict=False,
+                disable_forced_specializations=False,
```
I don't think we ever want to make this public, and we might be able to get rid of this with hybrid symints. 

1. So it deserves at least a `_` prefix.
2. Maybe we should guard against its public use?

avikchaudhuri commented on torch/_export/serde/schema.py:
```diff
@@ -359,6 +359,7 @@ class GraphModule:
     # the modules in order to unflatten the modules back to the eager calling
     # conventions.
     module_call_graph: List[ModuleCallEntry]
+    disable_forced_specializations: bool = False
```
This feels very wrong. I don't think we should get rid of checks here, but rather try converting them to real checks (so IR doesn't have this field). Or just not do any checks at all if this flag is `True`, in which rename the flag to `disable_runtime_assertions`.

avikchaudhuri commented on torch/_export/utils.py:
```diff
@@ -144,6 +147,10 @@ def get_keystr(key_path: KeyPath) -> str:
                                 )
                 else:
                     if arg_dim != node_dim:
+                        if disable_forced_specializations and isinstance(
```
So this is the case where multiple symbols occur, yes? Do we need a flag here? Maybe it's better to always ignore, or always try harder and always convert them to checks.

pianpwk replied:
Yep that's correct. Making a note here about maybe using these as guards and substituting in concrete values.

avikchaudhuri reviewed (changes requested):
Requesting changes mostly because I think they need some more discussion  (see inline comments).

avikchaudhuri commented on torch/export/_trace.py:
```diff
@@ -928,6 +937,12 @@ def _export(
+    if _disable_forced_specializations and strict:
+        raise UserError(
+            UserErrorType.INVALID_INPUT,
+            "_disable_forced_specializations can be only be specified in non-strict mode.",
```
Why?

pianpwk replied:
It's really messy to try to pass the flag into Dynamo's call to produce_guards(): https://github.com/pytorch/pytorch/blob/9aa7699185e4ec39077e3046dfd63244dffa9ddb/torch/_dynamo/guards.py#L1447

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1881,14 +1885,16 @@ def remove_redundant_dynamic_results(self):
     def prettify_results(
         self,
         original_signature: inspect.Signature,
+        dynamic_shapes=None,
```
How does this help with determinism?

pianpwk replied:
Sorry this is unclean, this is left over from trying to order results with the position in the shapes spec. No longer doing that.

avikchaudhuri commented on torch/export/_trace.py:
```diff
@@ -1260,6 +1278,7 @@ def forward(self, *args, **kwargs):
         if free_unbacked_symbols(k)
     }
     gm.meta["forward_arg_names"] = forward_arg_names
+    gm.meta["disable_forced_specializations"] = _disable_forced_specializations
```
Where's this used? Also this is strict, yeah? So redundant with your current assumption that strict is excluded.

avikchaudhuri commented on torch/export/_trace.py:
```diff
@@ -1071,6 +1088,7 @@ def forward(self, *args, **kwargs):
         assert out_spec is not None
 
         gm = ep_non_strict.gm
+        gm.meta["disable_forced_specializations"] = _disable_forced_specializations
```
Where's this used? Is this left over from a previous commit?

pianpwk replied:
Yeah, sorry, need to clean this up too.

avikchaudhuri reviewed (approved):
Some more questions. Otherwise looks OK.

pianpwk committed:
[export] disable_forced_specializations (#124949)

Summary:
By default, some inferred dynamic shapes guards/constraints that are not expressible with the current dynamic shapes language will lead to specialization to the concrete input values provided. If disable_forced_specializations is set to True, we will not specialize, and will not perform runtime checks on such produced guards. Instead, we allow the user to specify arbitrary shapes, and fail during runtime if the inputs are invalid. Constraints expressible with the language (e.g. ranges, linear derived dims) will still be enforced, and behavior for all other guards remains the same.

Cases where we typically specialize are reshapes:
```
x: [4, 6]  # [s0, s1]
x = x.reshape([x.shape[0] - 1, -1])
# this emits a guard Mod(s0*s1, s0-1) = 0, we specialize on s0=4, s1=6

x: [4, 6], y: [24]  # [s0, s1], [s2]
x = x.reshape([-1]) + y
# this emits a guard s0*s1 = s2, we specialize on s0=4, s1=6, s2=24
```

For now only applicable for non-strict mode (need to figure out how to pass this flag into dynamo's call of produce_guards).


Test Plan:
Added test case that checks compilation, runtime, and suggested fixes behavior.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

Reviewed By: avikchaudhuri

Differential Revision: D56361177

Pulled By: pianpwk

zhxchen17 commented:
@pytorchbot merge
